common/setups/rasr/nn_system.py

```diff
@@ -0,0 +1,554 @@
+    - cv is dev for returnn training
+    - dev for lm param tuning
+    - test corpora for final eval
+
+    settings needed:
+    - am
+    - lm
+    - lexicon
+    - ce training
+    - ce recognition
+    - ce rescoring
+    - smbr training
+    - smbr recognition
+    - smbr rescoring
+    """
+
+    def __init__(
+        self,
+        returnn_root: Optional[str] = None,
+        returnn_python_home: Optional[str] = None,
+        returnn_python_exe: Optional[str] = None,
+    ):
+        super().__init__()
+
+        self.returnn_root = returnn_root or gs.RETURNN_ROOT
+        self.returnn_python_home = returnn_python_home or gs.RETURNN_PYTHON_HOME
+        self.return_python_exe = returnn_python_exe or gs.RETURNN_PYTHON_EXE
+
+        self.tf_fwd_input_name = "tf-fwd-input"
+
+        self.devtrain_corpora = []
+        self.cv_corpora = []
+
+        self.train_input_data = None
+        self.cv_input_data = None
+        self.devtrain_input_data = None
+        self.dev_input_data = None
+        self.test_input_data = None
+
+        self.train_cv_pairing = None
+
+        self.datasets = {}
+
+        self.oggzips = {}
+
+        self.nn_configs = {}
+        self.nn_models = {}
+
+    # -------------------- Helpers --------------------
+    @staticmethod
+    def adapt_returnn_config_for_recog(returnn_config: returnn.ReturnnConfig):
+        """
+        Adapt a RETURNN config for recognition, e.g., remove loss and use log softmax activation in last layer
+
+        :param ReturnnConfig config:
+        :rtype ReturnnConfig:
+        """
+        assert isinstance(returnn_config, returnn.ReturnnConfig)
+        config = copy.deepcopy(returnn_config)
+        forward_output_layer = config.config.get("forward_output_layer", "output")
+        network = config.config.get("network")
+        for layer_name, layer in network.items():
+            if layer.get("unit", None) in {"lstmp"}:
+                layer["unit"] = "nativelstm2"
+            if layer.get("target", None) and layer_name != forward_output_layer:
+                layer.pop("target")
+                layer.pop("loss", None)
+                layer.pop("loss_scale", None)
+        network[forward_output_layer]["class"] = "linear"
+        network[forward_output_layer]["activation"] = "log_softmax"
+        network[forward_output_layer].pop("target", None)
+        # target = 'classes'
+        if "cropped" in network:
+            if network["output"]["from"] == ["cropped"]:
+                network["output"]["from"] = "upsample"
+            network.pop("cropped")
+            if "lstm_bwd_1" in network:
+                network["lstm_bwd_1"]["from"] = "upsample"
+                network["lstm_fwd_1"]["from"] = "upsample"
+            if "lstm_fwd_1_no_init" in network:
+                network["lstm_bwd_1_no_init"]["from"] = "upsample"
+                network["lstm_fwd_1_no_init"]["from"] = "upsample"
+
+        return config
+
+    @staticmethod
```

Comment:
I think this helper can be external at some point. Or actually I see a lot of overlap potential in my current CTCSystem vs the NNSystem here. I think it would be nice if the CTCSystem can fully extend from it. Then this would be fine here, but still people might want to build flows without the System class. All other flows are also defined standalone. (But there is also no problem in moving this later)

Reply:
Yes, we should maybe discuss this on Tuesday, how CTCSystem and NNSystem are going to relate to each other.

maybe a `flow` dir in `setups`??

Reply:
maybe this could be in i6_core/returnn/flow.py or so...? I am not sure.